File: sol/959.py
class Solution:
    def regionsBySlashes(self, grid: List[str]) -> int:
        n = len(grid)
        visited = [False] * (n * n * 4)
        res = 0
        def mark(i, j, k, coor):
            nonlocal visited
            if visited[coor]:
                return
            visited[coor] = True
            if grid[i][j] != '/':
                mark(i, j, 3 - k, coor - k * 2 + 3)
            if grid[i][j] != '\\':
                mark(i, j, k ^ 1, coor ^ 1)
            if k == 0 and j != n - 1:
                mark(i, j + 1, 2, coor + 6)
            elif k == 1 and i != n - 1:
                mark(i + 1, j, 3, coor + n * 4 + 2)
            elif k == 2 and j != 0:
                mark(i, j - 1, 0, coor - 6)
            elif k == 3 and i != 0:
                mark(i - 1, j, 1, coor - n * 4 - 2)
        coor = 0
        for i in range(n):
            for j in range(n):
                for k in range(4):
                    if not visited[coor]:
                        res += 1
                        mark(i, j, k, coor)
                    coor += 1
        return res

File: sol/959/959.cpp
struct Solution {
    void mark(int i, int j, int k, int coor, const vector<string> &grid, vector<bool> &visited){
        if (visited[coor]){
            return;
        }
        visited[coor] = true;
        if (grid[i][j] != '/'){
            this->mark(i, j, 3 - k, coor - k * 2 + 3, grid, visited);
        }
        if (grid[i][j] != '\\'){
            this->mark(i, j, k ^ 1, coor ^ 1, grid, visited);
        }
        switch (k){
            case 0:
                if (j != grid.size() - 1){
                    this->mark(i, j + 1, 2, coor + 6, grid, visited);
                }
                break;
            case 1:
                if (i != grid.size() - 1){
                    this->mark(i + 1, j, 3, coor + grid.size() * 4 + 2, grid, visited);
                }
                break;
            case 2:
                if (j != 0){
                    this->mark(i, j - 1, 0, coor - 6, grid, visited);
                }
                break;
            case 3:
                if (i != 0){
                    this->mark(i - 1, j, 1, coor - grid.size() * 4 - 2, grid, visited);
                }
                break;
        }
    }
    int regionsBySlashes(vector<string>& grid) {
        int n = grid.size();
        vector<bool> visited(n * n * 4, false);
        int res = 0, coor = 0;
        for (int i = 0; i < n; ++i){
            for (int j = 0; j < n; ++j){
                for (int k = 0; k < 4; ++k){
                    if (!visited[coor]){
                        ++res;
                        this->mark(i, j, k, coor, grid, visited);
                    }
                    ++coor;
                }
            }
        }
        return res;
    }
};
